src/providers/deepinfra/chatComplete.ts

```diff
@@ -0,0 +1,176 @@
+import { DEEPINFRA } from "../../globals";
+import {
+  ChatCompletionResponse,
+  ErrorResponse,
+  ProviderConfig,
+} from "../types";
+
+// TODOS: this configuration does not enforce the maximum token limit for the input parameter. If you want to enforce this, you might need to add a custom validation function or a max property to the ParameterConfig interface, and then use it in the input configuration. However, this might be complex because the token count is not a simple length check, but depends on the specific tokenization method used by the model.
+// TODOS: this configuration might have to check on the max value of n
+
+export const DeepInfraChatCompleteConfig: ProviderConfig = {
+  model: {
+    param: "model",
+    required: true,
+    default: "deepinfra/airoboros-70b",
+  },
+  messages: {
+    param: "messages",
+    required: true,
+    default: [],
+  },
+  frequency_penalty: {
+    param: "frequency_penalty",
+    default: 0,
+    min: -2,
+    max: 2,
+  },
+  max_tokens: {
+    param: "max_tokens",
+    default: 100,
+    min: 1,
+  },
+  n: {
+    param: "n",
+    default: 1,
+    min: 1,
+    max: 1,
+  },
+  presence_penalty: {
+    param: "presence_penalty",
+    min: -2,
+    max: 2,
+    default: 0,
+  },
+  temperature: {
+    param: "temperature",
+    default: 1,
+    min: 0,
+    max: 2,
+  },
+  top_p: {
+    param: "top_p",
+    default: 1,
+    min: 0,
+    max: 1,
+  },
+  stop: {
+    param: "stop",
+    default: null,
+  },
+  stream: {
+    param: "stream",
+    default: false,
+  },
+};
+
+interface DeepInfraChatCompleteResponse extends ChatCompletionResponse {
+  id: string;
+  object: string;
+  created: number;
+  model: string;
+  usage: {
+    prompt_tokens: number;
+    completion_tokens: number;
+    total_tokens: number;
+  };
+}
+
+export interface DeepInfraErrorResponse {
+  message: string;
+}
+
+interface DeepInfraStreamChunk {
+  id: string;
+  object: string;
+  created: number;
+  model: string;
+  choices: {
+    delta: {
+      role?: string | null;
+      content?: string;
+    };
+    index: number;
+    finish_reason: string | null;
+  }[];
+}
+
+export const DeepInfraChatCompleteResponseTransform: (
```

Comment:
Error structure for deepinfra is different than OpenAI. You can replicate the error message by sending invalid fields to deepinfra like max_tokens="abc" or something. It sends an array of failed validation like this

```json
{
    "detail": [
        {
            "loc": [
                "body",
                "stream"
            ],
            "msg": "value could not be parsed to a boolean",
            "type": "type_error.bool"
        },
        {
            "loc": [
                "body",
                "max_tokens"
            ],
            "msg": "value is not a valid integer",
            "type": "type_error.integer"
        }
    ]
}
```

In this function instead of checking `if ("message" in response && responseStatus !== 200)`, we can check for `if ("detail" in response && responseStatus !== 200 && response.detail.length)` and then map the first error from the detail array to gateway error structure. Something like this:

```
const firstError = response.detail[0];
const errorField = firstError.loc.join(".");
const errorMessage = firstError.msg
const errorType = firstError.type

return {
  error: {
    message: `${errorField}: ${errorMessage}`,
    type: errorType,
    param: null,
    code: null,
  },
  provider: DEEPINFRA,
} as ErrorResponse;
```

Reply:
Thank You for guiding me!! will check it out

Reply:
Suggesting something like this:
```
    let firstError: Record<string, any> | undefined;
    let errorField: string | null = null;
    let errorMessage: string | undefined;
    let errorType: string | null = null;
    if (Array.isArray(response.detail)) {
        firstError = response.detail[0];
        errorField = firstError.loc?.join(".") ?? "";
        errorMessage = firstError.msg;
        errorType = firstError.type;
    } else {
        errorMessage = response.detail;
        errorField = null;
        errorType = null;
    }

    return {
      error: {
        message: `${errorField ? `${errorField}: ` : ''}${errorMessage}`,
        type: errorType,
        param: null,
        code: null,
      },
      provider: DEEPINFRA,
    } as ErrorResponse;
  }
  ```